

# Project: Analysing posts on Hacker News to find out which posts and hours receive the highest comments on average

## The goal of this project is to analyse two specific types of posts on Hacker News website: ask posts and show posts to figure out which type of post and hours has the higher chance of receiving comments

In [4]:
# Read the hacker_news.csv file in as a list of lists
from csv import reader
opened_file = open('hacker_news.csv')
file = reader(opened_file)
hn = list(file)

# Display the first five rows of hn
print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


**Removing headers from a list of lists**

In [5]:
# Remove the header of the list
headers = hn[0]
print(headers)
hn = hn[1:]
print('\n')
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


**Extracting Ask HN and Show HN Posts**

In [6]:
# Create three empty lists
ask_posts = []
show_posts = []
other_posts = []

# Loop through each row in hn
for row in hn:
    title = row[1]
    lc_title = title.lower()
    if lc_title.startswith('ask hn'):
        ask_posts.append(row)
    elif lc_title.startswith('show hn'):
        show_posts.append(row)
    else: 
        other_posts.append(row)

# Check number of three lists
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


**Calculating the Average Number of Comments for Ask HN and Show HN Posts**

In [7]:
# Find average number of comments for Ask HN
total_ask_comments = 0

# Loop through ask posts
for row in ask_posts:
    num_comments = row[4]
    total_ask_comments += int(num_comments)
    
# Compute the average
avg_ask_comments = total_ask_comments / len(ask_posts)
print(avg_ask_comments)

14.038417431192661


In [8]:
# Find average number of comments for Show HN
total_show_comments = 0

# Loop through show posts
for row in show_posts:
    num_comments = row[4]
    total_show_comments += int(num_comments)
    
# Compute the average
avg_show_comments = total_show_comments / len(show_posts)
print(avg_show_comments)

10.31669535283993


Ask posts receive more comments on average than show posts. We'll focus our remaining analysis just on these posts.

**Finding the Amount of Ask Posts and Comments by Hour Created**

In [19]:
# Import the datetime module
import datetime as dt

# Create an empty list
result_list = []

# Loop over the ask posts list
for row in ask_posts:
    new_list = []
    new_list.append(row[6])   # First element from column 'created at'
    new_list.append(row[4])   # Second element from column 'num_comments'
    result_list.append(new_list) # Append the new list to result list

# Create two empty dictionaries
counts_by_hour = {}
comments_by_hour = {}

# Loop through result list
for row in result_list:
    date = row[0]
    num_comments = int(row[1])
    # Create a datetime object
    date_dt = dt.datetime.strptime(date, "%m/%d/%Y %H:%M")
                                   
    # Select the hour from the datetime object and convert to a string
    hour_dt = date_dt.strftime("%H")
    
    #Check if hour in counts_by_hour
    if hour_dt not in counts_by_hour:
        counts_by_hour[hour_dt] = 1
        comments_by_hour[hour_dt] = num_comments
    else:
        counts_by_hour[hour_dt] += 1
        comments_by_hour[hour_dt] += num_comments

# Print results
print(counts_by_hour)
print('\n')
print(comments_by_hour)

{'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58}


{'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}


**Calculating the Average Number of Comments for Ask HN Posts by Hour**

In [20]:
# Create an empty list
avg_by_hour = []

# Loop through counts_by_hour dictionary

for hour in counts_by_hour:
    avg_num = comments_by_hour[hour] / counts_by_hour[hour]  # Calculate average number
    avg_by_hour.append([hour, avg_num])    # Append the hour and average number to the list

# Print result
print(avg_by_hour)

[['09', 5.5777777777777775], ['13', 14.741176470588234], ['10', 13.440677966101696], ['14', 13.233644859813085], ['16', 16.796296296296298], ['23', 7.985294117647059], ['12', 9.41095890410959], ['17', 11.46], ['15', 38.5948275862069], ['21', 16.009174311926607], ['20', 21.525], ['02', 23.810344827586206], ['18', 13.20183486238532], ['03', 7.796296296296297], ['05', 10.08695652173913], ['19', 10.8], ['01', 11.383333333333333], ['22', 6.746478873239437], ['08', 10.25], ['04', 7.170212765957447], ['00', 8.127272727272727], ['06', 9.022727272727273], ['07', 7.852941176470588], ['11', 11.051724137931034]]


**Sorting and Printing Values from a List of Lists**

In [40]:
# Create a list that equals avg_by_hour with swapped columns
swap_avg_by_hour = []

# Loop through avg_by_hour list
for row in avg_by_hour:
    hour = row[0]
    avg_num = row[1]
    swap_avg_by_hour.append([avg_num, hour]) # Swap 1st and 2nd element of the avg_by_hour list

# Print swap_avg_by_hour
print(swap_avg_by_hour)

# Sort the list in descending order
sorted_swap = sorted(swap_avg_by_hour,reverse=True)

[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]


In [41]:
# Print the string "Top 5 Hours for Ask Posts Comments"
print("Top 5 Hours for Ask Posts Comments ")
      
# Loop through sorted_swap
for row in sorted_swap[:5]:
    num_comments = row[0]
    hour = row[1]
    hour_dt = dt.datetime.strptime(hour, "%H")
    hour_formatted = hour_dt.strftime("%H:%M")
    
    str_result = "{}: {:.2f} average comments per post".format(hour_formatted, num_comments)
    print(str_result)

Top 5 Hours for Ask Posts Comments 
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The result shows that the hour that receives that most comments per post is 15:00, with an average of 38.59 comments. According to the [documentation for the dataset](https://www.kaggle.com/hacker-news/hacker-news-posts), the time zone is Eastern time in the US. The center of Australi is 14:45 hours ahead of the center of the US. Thus, generally, a post should be created at 5:45 AM in Australia to have a higher chance of receiving comments.

**Conclusion**

In this project, we analysed the ask posts and show posts on Hacker News to find out which type of posts and hours receive the most comments on average. The result is a post created at 5:45 AM in Australia named an ask post has a higher chance of receiving comments.